src/Servers/Kestrel/Core/src/Internal/Http/MessageBody.cs

```diff
@@ -165,16 +166,82 @@ protected ValueTask<ReadResult> StartTimingReadAsync(ValueTask<ReadResult> readA
-            _context.TimeoutControl.BytesRead(bytesRead - _alreadyTimedBytes);
-            _alreadyTimedBytes = 0;
+            var numFirstSeenBytes = bytesInReadResult - _alreadyTimedBytes;
+
+            if (numFirstSeenBytes > 0)
+            {
+                _context.TimeoutControl.BytesRead(numFirstSeenBytes);
+            }
+        }
+
+        protected void StopTimingRead(long bytesInReadResult)
+        {
+            CountBytesRead(bytesInReadResult);
 
             if (_backpressure)
             {
                 _backpressure = false;
                 _context.TimeoutControl.StopTimingRead();
             }
         }
+
+        protected long OnAdvance(ReadResult readResult, SequencePosition consumed, SequencePosition examined)
+        {
+            // This code path is fairly hard to understand so let's break it down with an example
+            // ReadAsync returns a ReadResult of length 50.
+            // Advance(25, 40). The examined length would be 40 and consumed length would be 25.
+            // _totalExaminedInPreviousReadResult starts at 0. newlyExamined is 40.
+            // OnDataRead is called with length 40.
+            // _totalExaminedInPreviousReadResult is now 40 - 25 = 15.
+
+            // The next call to ReadAsync returns 50 again
+            // Advance(5, 5) is called
+            // newlyExamined is 5 - 15, or -10.
+            // Update _totalExaminedInPreviousReadResult to 10 as we consumed 5.
+
+            // The next call to ReadAsync returns 50 again
+            // _totalExaminedInPreviousReadResult is 10
+            // Advance(50, 50) is called
+            // newlyExamined = 50 - 10 = 40
+            // _totalExaminedInPreviousReadResult is now 50
+            // _totalExaminedInPreviousReadResult is finally 0 after subtracting consumedLength.
+
+            long examinedLength, consumedLength, totalLength;
+
+            if (consumed.Equals(examined))
+            {
+                examinedLength = readResult.Buffer.Slice(readResult.Buffer.Start, examined).Length;
+                consumedLength = examinedLength;
+            }
+            else
+            {
+                consumedLength = readResult.Buffer.Slice(readResult.Buffer.Start, consumed).Length;
+                examinedLength = consumedLength + readResult.Buffer.Slice(consumed, examined).Length;
+            }
+
+            if (examined.Equals(readResult.Buffer.End))
+            {
+                totalLength = examinedLength;
+            }
+            else
+            {
+                totalLength = readResult.Buffer.Length;
+            }
+
+            var newlyExamined = examinedLength - _examinedUnconsumedBytes;
+
+            if (newlyExamined > 0)
+            {
+                OnDataRead(newlyExamined);
+                _examinedUnconsumedBytes += newlyExamined;
+            }
+
+            _examinedUnconsumedBytes -= consumedLength;
+            _alreadyTimedBytes = totalLength - consumedLength;
```

Comment:
I don't quite follow why _alreadyTimedBytes should be totalLength - consumedLength. Can you explain that?

Reply:
For timing purposes, we count all bytes that were provided by the ReadResult right when ReadAsync/TryRead returns and before Advance is called. This is before we know how much the consumer examined, and the caller might not examine everything.